gitoid/src/gitoid.rs

```diff
@@ -181,30 +175,36 @@ impl TryFrom<Url> for GitOid {
         };
 
         // Parse the hash algorithm, if present.
-        let hash_algorithm = {
-            let part = segments
-                .next()
-                .and_then(|p| p.is_empty().not().then_some(p))
-                .ok_or_else(|| MissingHashAlgorithm(url.clone()))?;
+        let hash_algorithm = segments
+            .next()
+            .and_then(|p| p.is_empty().not().then_some(p))
+            .ok_or_else(|| MissingHashAlgorithm(url.clone()))?;
 
-            HashAlgorithm::from_str(part)?
-        };
+        if hash_algorithm != D::NAME {
+            return Err(Error::MismatchedHashAlgorithm(
+                D::NAME.to_string(),
+                hash_algorithm.to_string(),
+            ));
+        }
 
         // Parse the hash, if present.
         let hex_str = segments
             .next()
             .and_then(|p| p.is_empty().not().then_some(p))
             .ok_or_else(|| MissingHash(url.clone()))?;
-        let mut value = [0u8; 32];
+
+        let mut value = GenericArray::generate(|_| 0);
```

Comment:
We have the match the `generic-array` version used in the crypto crates (`sha1`, `sha2`, `digest`), which is _not_ current (it's 0.14.x, instead of 1.x.y). This means we don't get the nice `arr` macro that `generic-array` 1.0.0 defines, and instead have to use `generate` here.

Reply:
I could probably put a `TODO` here.